packages/@lwc/engine-core/src/framework/vm.ts

```diff
@@ -749,6 +764,30 @@ export function resetComponentRoot(vm: VM) {
     vm.velements = EmptyArray;
 }
 
+// Helper function to traverse a tree of VFragment nodes and remove all root children.
+// This is moved into a separate function to minimize the perf/mem impact of the stack traversal
+// on non-vfragment use cases
+function removeFragmentChildren(vnode: VFragment, vm: VM) {
+    const {
+        renderRoot,
+        renderer: { remove },
+    } = vm;
+
+    const nodeStack: VNode[] = [];
+    ArrayPush.call(nodeStack, ...vnode.children);
+
+    let currentNode: VNode | null | undefined;
+    while (!isUndefined((currentNode = ArrayPop.call(nodeStack)))) {
+        if (!isNull(currentNode)) {
+            if (isVFragment(currentNode)) {
+                ArrayPush.call(nodeStack, ...currentNode.children);
```

Comment:
The order of removal here is not following tree order. Not a problem now, because disconnectedCallback is being managed manually by `runChildNodesDisconnectedCallback`. When the switch is made to native disconnectedCallback for custom elements, this will need to be fixed.

fyi @nolanlawson

Reply:
Thanks for calling that out. We should change it to match the disconnect order.

Reply:
Good catch @ravijayaramappa !

Reply:
@ravijayaramappa Looked into it a bit, you're right in that there's an issue with disconnectedCallback ordering, but not in the way you described; the order of removal here actually does match the order in `runChildNodesDisconnectedCallback`; both remove/disconnect in reverse sibling order.

However, other issues have come up, and I'm hitting up against the inconsistent disconnect ordering mentioned in https://github.com/salesforce/lwc/issues/2713. When you have some time, could you take a look at https://github.com/salesforce/lwc/pull/3369 and give your input on what the "correct" order of removal should be in the various use cases?